build.yaml

```diff
@@ -694,13 +811,124 @@ common_cli:
+
+    # vagrant doesn't support fully semantic versioning, numbering must be x.y.z
+    # https://github.com/mitchellh/vagrant/issues/6052
+    VAGRANT_BOX_VERSION="\$( echo \${FLOCKER_VERSION} | cut -f 1,2,3 -d "." )"
+
+    cd \$BOX_PATH
+    # generates the metadata.json, it used by vagrant box update operations.
+    cat <<EOF_METADATA > metadata.json
+    {
+      "name": "clusterhq/flocker-tutorial",
+      "description": "This box contains the flocker-tutorial VM.",
+      "versions": [{
+        "version": "\${VAGRANT_BOX_VERSION}",
+        "providers": [{
+          "name": "virtualbox",
+          "url": "\${HTTP_BOX_URL}",
+          "checksum_type": "sha1",
+          "checksum": "\${SHA1}"
+        }]
+      }]
+    }
+    EOF_METADATA
+
+    cp metadata.json \${BOX_NAME}.json
+
+    # uploads the vagrant box and metadata files to S3
+    # We won't overwrite the json metadata files if the upload of the box
+    # failed
+    box_files="\${BOX_NAME} \${BOX_NAME}.json metadata.json"
+    (
+      s3_upload \
+        \${S3_BUCKET}/\${BOX_PATH}/ "\${BOX_NAME}" && \
+      s3_upload \
+        \${S3_BUCKET}/\${BOX_PATH}/ "\${BOX_NAME}.json" && \
+      s3_upload \
+        \${S3_BUCKET}/\${BOX_PATH}/ "metadata.json"
+    ) ||  abort_build
+
+    cd -
+    echo "This vagrant box is available at : \${HTTP_BOX_JSON_LATEST}"
+
+  upload_vagrant_basebox_osx_yosemite_to_s3: &upload_vagrant_basebox_osx_yosemite_to_s3 |
+    # uploads the generated vagrant box to a folder on the S3
+    # clusterhq-vagrant-jenkins-boxes bucket.
+    BOX_NAME="jenkins-osx-yosemite-\${FLOCKER_VERSION}.box"
+    SHA1=\$( sha1sum \${BOX_NAME} | cut -f 1 -d " " )
+
+    S3_BOX_URL=s3://\${S3_BUCKET}/\${BOX_NAME}
+
+    # the JSON metadata for our box is jenkins-osx-yosemite-version.box.json
+    S3_BOX_JSON=s3://\${S3_BUCKET}/\${BOX_PATH}/\${BOX_NAME}.json
+
+    # the JSON url below is what the vagrant file will consume
+    S3_BOX_JSON_LATEST=s3://\${S3_BUCKET}/\${BOX_PATH}/metadata.json
+
+    # vagrant doesn't support fully semantic versioning, numbering must be x.y.z
+    # https://github.com/mitchellh/vagrant/issues/6052
+    VAGRANT_BOX_VERSION="\$( date +%s).0.0"
+
+    # generates the metadata.json, it used by vagrant box update operations.
+    create_vagrant_metadata_file \
+      metadata.json \
+      clusterhq/jenkins-osx-yosemite \
+      'This box contains the OSX box used for the client tests.' \
+      \${VAGRANT_BOX_VERSION} \
+      \${S3_BOX_URL} \
+      \${SHA1}
+
+    cp metadata.json \${BOX_NAME}.json
+
+    # uploads the vagrant box and metadata.json files
+    # We won't overwrite the json metadata files if the upload of the box
+    # failed
+    (
+      s3_upload \${S3_BUCKET} "\${BOX_NAME}" &&  \
+      s3_upload \${S3_BUCKET} "\${BOX_NAME}.json" &&  \
+      s3_upload \${S3_BUCKET} "metadata.json"
+    ) || abort_build
+
+    echo "This vagrant box is available is available at :"
+    echo "\${S3_BOX_JSON_LATEST}"
```

Comment:
:question: These two stanzas (`upload_vagrant_basebox_osx_yosemite_to_s3` and `upload_vagrant_tutorial_basebox_to_s3`) are really similar. Any way to share more code?